README.md
# Biblioteca de Tradução

[![N|Solid](https://img.shields.io/badge/por-%40ronnyjss-RONNY?logo=xcode&logoColor=%23FFFFFF&style=flat-square&color=00ABFF)](https://github.com/ronnyjss)
[![N|Solid](https://img.shields.io/badge/PHP-%5E8.0-RONNY?logo=php&logoColor=%23FFFFFF&style=flat-square&color=00ABFF)](https://www.php.net)
[![N|Solid](https://img.shields.io/github/license/ronnyjss/traducao?logo=github&logoColor=%23FFFFFF&style=flat-square&color=%2300ABFF&label=Licença)](LICENSE)

[![N|Solid](https://img.shields.io/badge/Packagist-ronnyjss/traducao-RONNY?logo=packagist&logoColor=%23FFFFFF&style=flat-square&color=00ABFF)](https://packagist.org/packages/ronnyjss/traducao)
[![N|Solid](https://img.shields.io/github/v/tag/ronnyjss/traducao.svg?color=00ABFF&label=Vers%C3%A3o&logo=github&logoColor=%23FFFFFF&sort=semver&style=flat-square)](https://github.com/ronnyjss/traducao/releases)
[![N|Solid](https://img.shields.io/github/workflow/status/ronnyjss/traducao/Teste?label=Teste&logo=githubactions&logoColor=FFFFFF&style=flat-square)](https://github.com/ronnyjss/traducao/actions)

Criado por Ronny Santos ronny@jss.art.br (MIT License)

Auxilia na importação e utilização de gettext em arquivos PO.
Extendendo da Lib: [gettext/gettext](https://packagist.org/packages/gettext/gettext).

Todas as mudanças notáveis neste projeto serão documentadas no [Log de Alterações](CHANGELOG.md), nele estara presente as informações sobre mudanças recentes. O formato é baseado em [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

Este projeto adere ao [Controle de Versão Semântica](https://semver.org/spec/v2.0.0.html).

## Instação

```
composer require ronnyjss/traducao
```

## Exemplo de Uso

Pasta do [Exemplo Funcional](exemplo).

```php
use ronnyjss\Traducao;

/**
* Informa ao sistema qual a linguagem a ser usada.
*
* 'English' - Usualmente é como o pacote de linguagem Inglês(Estados Unidos),
* é encontrado na plataforma Windows.
*
* 'en_US' - É a forma mais conhecida do pacote de linguagem Inglês(Estados Unidos),
* e como ele é encontrado na plataformas Linux.
*
* 'en' - É a forma generica do pacote Inglês internacional.
* Usado para casos em que o pacote especifico não é encontrado,
* então o pacote internacional é selecionado.
**/
setlocale(LC_ALL_'en_US', 'en');

/**
* Informa a biblioteca a onde se encontra a pasta com os arquivos de linguagens disponiveis.
*
* Com base na linguagem setada a cima,
* a lib consultara essa pasta em busca dos seguintes arquivos:
* 'en_US.po' e 'en.po'
*
* Caso no setlocale, tenha sido informado a codificação,
* ela tambem ira buscar mais dois arquivos adicionais.
* Ex: setlocale('English.utf8', 'en_US.utf8', 'en.utf8');
* Ira buscar os arquivos nessa sequencia, o primeiro encontrado será utilizado:
* 'en_US.utf8.po', 'en_US.po', 'en.utf8.po' e 'en.po'.
**/
Traducao::adicionarPO(__DIR__ . '/linguagem');

/**
* Função responsavel por substituir a frase,
* pela respectiva tradução na linguagem atual.
*
* No primeiro parametro é informado a string que deve ser traduzida.
* No segundo parametro é informado o dominio que incapsula o pacote da linguagem.
*
* Neste exemplo abaixo, ela irá retornar a frase:
* 'Hello visitor, welcome to our system.'
**/
echo Traducao::traduzir('Ola Visitante, Bem vindo ao nosso sistema.', 'ronnyjssTraducaoExemplo');

/**
* Recomendo altamente que seja criada uma função substituta do metodo nativo da lib,
* para um codigo mais limpo. ficando algo do tipo:
*
* @param string	$s_mensagem
* @param array	$a_opcao
* @return string
**/
function __(
	string $s_mensagem,
	string $a_opcao = 'ronnyjssTraducaoExemplo'
):string {
	return Traducao::traduzir($s_mensagem, $a_opcao);
}

# Utilizando a função, o exemplo de echo acima ficaria assim:
echo __('Ola Visitante, Bem vindo ao nosso sistema.');
```

## Informações

Por favor, veja o [Log de Alterações](CHANGELOG.md) para mais informações sobre mudanças recentes.

Sob a Licença do MIT. Consulte [Licença](LICENSE) para obter mais informações.


CHANGELOG.md
# Log de Alterações

Todas as mudanças notáveis neste projeto serão documentadas neste arquivo.

O formato é baseado em [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
e este projeto adere ao [Controle de Versão Semântica](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2021-11-30

### Adicionado

- Criado a classe `ronnyjss/Traducao`.
- Adicionado à classe, o metodo `Traducao::adicionarPo` e `Traducao::traduzir`.
- Criado a função pra `traducao\__`, que extende do metodo `Traducao::traduzir`. Função criada pra proporcionar um codigo mais limpo, e para que os arquivos tenha suporte aos geradores de tradução automatica baseados em Wordpress.

## [1.0.1] - 2021-11-30

### Adicionado

- Adicionado o Badge referente ao pacote no Packist.

### Corrigido

- Badge `Teste` e `Versão`
- Nomenclatura do Badge Licença
- Workflow de Testes
- Composer.json

## [1.0.2] - 2021-12-03

### Adicionado

- Adicionado novos arquivos de linguagem para teste.
- Adicionado novos testes baseados em etapas 1-6;

### Atualizado

- Listagem de linguagens e regiões no arquivo de constantes.
- Testes, que eram feitos de maneira rasa, agora feitos de maneira mais ampla.
- Forma de busca pelos arquivos de linguagem.

## Removido
- Testes imprecisos.


tests/TraducaoTest.php
<?php namespace ronnyjss\traducao\teste;

use	Exception,
	ronnyjss\Traducao,
	PHPUnit\Framework\TestCase;

/**
* Testr da classe Traducao.
*
* @package		ronnyjss\traducao\teste
* @author		Ronny Santos <ronny@jss.art.br>
* @copyright	© 2021 RonnyJSS
* @license		AGPL-3.0
* @link			github.com/ronnyjss/traducao
**/
class TraducaoTest extends TestCase
{

	# ID da frase de Teste.
	const FRASE_ID = 'linguagem.atual';

	# Dominio de teste padrão, utilizado para testar a tradução.
	const DOMINIO = 'ronnyjssTraducao';

	/**
	* Função de teste padrão.
	*
	* @return void
	**/
	private function executarTestePadrao(
		string	$s_linguagem,
		string	$s_retornoEsperado
	):void {
		$a_linguagem[0] = LC_ALL;
		$a_linguagem[1] = $s_linguagem;
		if((bool) strpos($s_linguagem, '_')) $a_linguagem[2] = str_replace('_', '-', $s_linguagem);

		call_user_func_array('setlocale', $a_linguagem);

		Traducao::adicionarPO(dirname(__DIR__) . '/src/linguagem');

		$this->assertEquals(
			$s_retornoEsperado,
			Traducao::traduzir(self::FRASE_ID, self::DOMINIO)
		);
	}

	/**
	* Teste de Exception para linguagem indisponível.
	*
	* @return void
	**/
    public function testTraducaoException(
	) {
		setlocale(LC_ALL, 'uk_UA', 'uk-UA', 'uk');

		$this->expectException(Exception::class);
		Traducao::adicionarPO(dirname(__DIR__) . '/src/linguagem');
		$this->expectExceptionMessage('Nenhum arquivo na linguagem uk foi encontrado.');
	}

	/**
	* Teste de tradução da Etapa de busca 1.
	*
	* Entrada: pt_PT.utf8
	* Retorno: pt_PT.utf8
	*
	* @return void
	**/
	public function testTraducaoEtapa1(
	) {
		$this->executarTestePadrao(
			'pt_PT.utf8',
			'A linguagem atual é pt_PT.utf8.'
		);
	}

	/**
	* Teste de tradução da Etapa de busca 2.
	*
	* Entrada: en_US
	* Retorno: en_US
	*
	* Entrada: en_US.utf8
	* Retorno: en_US
	*
	* @return void
	**/
	public function testTraducaoEtapa2(
	) {
		$this->executarTestePadrao(
			'en_US',
			'The current language is en_US.'
		);

		$this->executarTestePadrao(
			'en_US.utf8',
			'The current language is en_US.'
		);
	}

	/**
	* Teste de tradução da Etapa de busca 3.
	*
	* Entrada: zh_CN.utf8
	* Retorno: zh.utf8
	*
	* @return void
	**/
	public function testTraducaoEtapa3(
	) {
		$this->executarTestePadrao(
			'zh_CN.utf8',
			'当前语言是ZH.UTF8。'
		);
	}

	/**
	* Teste de tradução da Etapa de busca 4.
	*
	* Entrada: es_ES
	* Retorno: es
	*
	* Entrada: es_ES.utf8
	* Retorno: es
	*
	* @return void
	**/
	public function testTraducaoEtapa4(
	) {
		$this->executarTestePadrao(
			'es_ES',
			'El idioma actual es ES.'
		);

		$this->executarTestePadrao(
			'es_ES.utf8',
			'El idioma actual es ES.'
		);
	}

	/**
	* Teste de tradução da Etapa de busca 5.
	*
	* Entrada: ar_AE.utf8
	* Retorno: ar_EG.utf8
	*
	* @return void
	**/
	public function testTraducaoEtapa5(
	) {
		$this->executarTestePadrao(
			'ar_AE.utf8',
			'اللغة الحالية هي ar_EG.'
		);
	}

	/**
	* Teste de tradução da Etapa de busca 6.
	*
	* Entrada: pt_PT
	* Retorno: pt_BR
	*
	* Entrada: de_LU.utf8
	* Retorno: de_CH
	*
	* @return void
	**/
	public function testTraducaoEtapa6(
	) {
		$this->executarTestePadrao(
			'pt_PT',
			'A linguagem atual é pt_BR.'
		);

		$this->executarTestePadrao(
			'de_LU.utf8',
			'Die aktuelle Sprache ist de_CH.'
		);
	}

	/**
	* Testa o uso de pluralidade de palavras.
	*
	* @return void
	**/
	public function testTraducaoPlural(
	) {
		setlocale(LC_ALL, 'en_US', 'en-US', 'en');

		Traducao::adicionarPO(dirname(__DIR__) . '/src/linguagem');

		$this->assertEquals(
			'A file has been removed.',
			Traducao::traduzir(
				'Um arquivo foi removido.',
				self::DOMINIO
			)
		);

		$this->assertEquals(
			'6 files have been removed.',
			sprintf(Traducao::traduzir(
				'%s arquivos foram removidos.',
				self::DOMINIO
			), 6)
		);
	}
}

/**
* en_US
* -----
* en_US			-> Etapa 2
* en			-> Etapa 4
* en_*			-> Etapa 6
*
*
* en_US.utf8
* -----
* en_US.utf8	-> Etapa 1
* en_US			-> Etapa 2
* en.utf8		-> Etapa 3
* en			-> Etapa 4
* en_*.utf8		-> Etapa 5
* en_*			-> Etapa 6
*
*
* en
* -----
* en			-> Etapa 4
* en_*			-> Etapa 6
*
*
* en.utf8
* -----
* en.utf8		-> Etapa 3
* en_*.utf8		-> Etapa 5
* en_*			-> Etapa 6
**/

src/funcao.php
<?php
/**
* Biblioteca para importar gettext de arquivos PO.
* PHP ^8.0
*
* @package		ronnyjss\traducao
* @author		Ronny Santos <ronny@jss.art.br>
* @copyright	© 2021 RonnyJSS
* @license		AGPL-3.0
* @link			github.com/ronnyjss/traducao
**/

namespace ronnyjss\traducao;

use ronnyjss\Traducao;

/**
* Função para facil utilização da classe de tradução.
*
* @param string $s_mensagem
* @param array $a_opcao
* @return string
**/
function __(
	string $s_mensagem,
	string $a_opcao = 'ronnyjssTraducao'
):string {
	return Traducao::traduzir($s_mensagem, $a_opcao);
}

/**
* Recebe o caminho de uma pasta ou arquivo,
* e retorna o mesmo com o separador de arquivo
* utilizado pelo sistema operacional atual.
*
* @param string $s_caminho
* @return string
**/
function sd(
	string $s_caminho
):string {
	return str_replace(['/', '\\'], DIRECTORY_SEPARATOR, $s_caminho);
}

src/Traducao.php
<?php namespace ronnyjss;

use Exception, Gettext\Loader\PoLoader;

use function ronnyjss\traducao\{
	__,
	sd
};

use const ronnyjss\traducao\LINGUAGENS;

/**
* Classe para importar e utilizar gettext de arquivos PO.
*
* @package		ronnyjss\traducao
* @author		Ronny Santos <ronny@jss.art.br>
* @copyright	© 2021 RonnyJSS
* @license		AGPL-3.0
* @link			github.com/ronnyjss/traducao
**/
class Traducao
{

	# Armazena a lista de linguagens carregadas.
	private static ?array $a_linguagens = [];

	/**
	* Recebe a pasta com as linguagens disponiveis,
	* e carrega a linguagem setada no sistema.
	*
	* @param string $s_pasta
	* @return void
	**/
	public static function adicionarPO(
		string $s_pasta
	):void {
		preg_match(
			'/(?<linguagem>[a-z]+)(?:[_-](?<regiao>[a-z]+))?(?:\.(?<codificacao>[a-z0-9]+))?/i',
			setlocale(LC_ALL, 0),
			$a_linguagem
		);

		$s_linguagem	= $a_linguagem['linguagem'] ?? null;
		$s_regiao		= $a_linguagem['regiao'] ?? null;
		$s_codificacao	= strtolower($a_linguagem['codificacao']?? null);

		unset($a_linguagem);

		if(!in_array($s_linguagem, array_keys(LINGUAGENS))) {
			throw new Exception(sprintf(__('Linguagem "%s" não suportada.'), $s_linguagem));
			$s_linguagem = 'en';
		}

		for($i_etapa = 1; $i_etapa <= 6; $i_etapa++) {
			if($i_etapa == 1 && !empty($s_regiao) && !empty($s_codificacao))
				$s_arquivo = "{$s_linguagem}_$s_regiao.$s_codificacao.po";

			if($i_etapa == 2 && !empty($s_regiao))
				$s_arquivo = "{$s_linguagem}_$s_regiao.po";

			if($i_etapa == 3 && !empty($s_codificacao))
				$s_arquivo = "$s_linguagem.$s_codificacao.po";

			if($i_etapa == 4)
				$s_arquivo = "$s_linguagem.po";

			if($i_etapa <= 4 && isset($s_arquivo))
				try {
					$o_linguagem = (new PoLoader())->loadFile(sd("$s_pasta/$s_arquivo"));
					break;
				} catch(Exception $o_exception) {}

			if($i_etapa == 5 && !empty($s_codificacao))
				foreach(LINGUAGENS[$s_linguagem] as $s_t_regiao)
					try {
						$o_linguagem = (new PoLoader())->loadFile(sd("$s_pasta/{$s_linguagem}_$s_t_regiao.$s_codificacao.po"));
						break 2;
					} catch(Exception $o_exception) {}

			if($i_etapa == 6)
				foreach(LINGUAGENS[$s_linguagem] as $s_t_regiao)
					try {
						$o_linguagem = (new PoLoader())->loadFile(sd("$s_pasta/{$s_linguagem}_$s_t_regiao.po"));
						break 2;
					} catch(Exception $o_exception) {}
		}

		if(isset($o_linguagem)) {
			if(method_exists($o_linguagem, 'getDomain'))
				self::$a_linguagens[$o_linguagem->getDomain() ?? 'default'] = $o_linguagem;
			else
				throw new Exception(__('O arquivo não é um arquivo de linguagem, ou não pode ser interpretado corretamente.'));
		} else
			throw new Exception(sprintf(__('Nenhum arquivo na linguagem %s foi encontrado.'), $s_linguagem));
	}

	public static function traduzir(
		string	$s_mensagem,
		string	$s_dominio = 'default',
	):string {
		if(!empty(self::$a_linguagens) && isset(self::$a_linguagens[$s_dominio]))
			foreach(self::$a_linguagens[$s_dominio]->getTranslations() AS $o_traducao)
				if($o_traducao->getOriginal() === $s_mensagem) {
					$s_mensagem = $o_traducao->getTranslation();
					break;
				} else if($o_traducao->getPlural() === $s_mensagem) {
					$s_mensagem = $o_traducao->getPluralTranslations()[0] ?? $o_traducao->getTranslation();
					break;
				}

		return $s_mensagem;
	}
}

exemplo/index.php
<?php namespace ronnyjss\traducao\exemplo;
/**
* Biblioteca para importar gettext de arquivos PO.
* PHP ^8.0
*
* @package		ronnyjss\traducao
* @author		Ronny Santos <ronny@jss.art.br>
* @copyright	© 2021 RonnyJSS
* @license		AGPL-3.0
* @link			github.com/ronnyjss/traducao
**/

use ronnyjss\Traducao;

require_once __DIR__ . '/vendor/autoload.php';

/**
* Informa ao sistema qual a linguagem a ser usada.
*
* 'en_US' - É a forma mais conhecida do pacote de linguagem Inglês(Estados Unidos),
* e como ele é encontrado na plataformas Linux.
*
* 'en-US' - Usualmente é como o pacote de linguagem Inglês(Estados Unidos),
* é encontrado na plataforma Windows.
*
* 'en' - Informa que a linguagem a ser usada pertence a lingua Inglêsa.
* Usado para casos em que o pacote especifico não é encontrado,
* então um pacote disponivel da lingua usual é selecionado.
**/
setlocale(LC_ALL, 'en_US', 'en-US', 'en');

/**
* Informa a biblioteca a onde se encontra a pasta com os arquivos de linguagens disponiveis.
*
* Com base na linguagem setada acima,
* a lib consultara essa pasta em busca dos seguintes arquivos:
*
* 1. Busca direta:
*	Irá buscar exatamente a linguagem com região informada, portando tentara encontrar o arquivo:
*		'en_US.po'
* 2. Busca recursiva:
*	Irá buscar todas as linguagens disponiveis na pasta, com base na linguagem usual,
*	portanto tentara encontrar os arquivos:
*		'en.po', 'en_US.po', 'en_AG.po', 'en_AU.po', 'en_BW.po', 'en_CA.po', 'en_DK.po', 'en_GB.po',
*		'en_HK.po', 'en_IE.po', 'en_IN.po', 'en_NG.po', 'en_NZ.po', 'en_PH.po', 'en_SG.po', 'en_ZA.po',
*		'en_ZM.po' e 'en_ZW.po'
*	O primeiro arquivo encontrado, sera usado.
*
* Caso a codificação da linguagem seja informada, exemplo:
*	'en_US.UTF-8'
*		Aplica-se os mesmos casos acima,
*		com a diferença que caso a busca direta ao arquivo 'en_US.utf8.po' falhe,
*		a busca direta ira fazer mais uma busca ao arquivo sem codificação 'en_US.po', caso não encontre também,
*		a busca recursiva irá buscar todos os arquivos com e sem a codificação 'utf8'.
**/
Traducao::adicionarPO(__DIR__ . '/linguagem');

/**
* Função responsavel por substituir a frase,
* pela respectiva tradução na linguagem atual.
*
* No primeiro parametro é informado a string que deve ser traduzida.
* No segundo parametro é informado o dominio que incapsula o pacote da linguagem.
*
* Neste exemplo abaixo, ela irá retornar a frase:
* 'Hello visitor, welcome to our system.'
**/
echo Traducao::traduzir('Ola Visitante, Bem vindo ao nosso sistema.', 'ronnyjssTraducaoExemplo');
echo '</br>';

/**
* Recomendo altamente que seja criada uma função substituta do metodo nativo da lib,
* para um codigo mais limpo. ficando algo do tipo:
*
* @param string	$s_mensagem
* @param array	$a_opcao
* @return string
**/
function __(
	string $s_mensagem,
	string $a_opcao = 'ronnyjssTraducaoExemplo'
):string {
	return Traducao::traduzir($s_mensagem, $a_opcao);
}

# Utilizando a função, o exemplo de echo acima ficaria assim:
echo __('Ola Visitante, Bem vindo ao nosso sistema.');